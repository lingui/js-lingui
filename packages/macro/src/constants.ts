export const ID = "id"
export const MESSAGE = "message"
export const COMMENT = "comment"
export const EXTRACT_MARK = "i18n"
export const CONTEXT = "context"
