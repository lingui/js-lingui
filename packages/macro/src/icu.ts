import {
  Expression,
  isJSXEmptyExpression,
  JSXElement,
  Node,
} from "@babel/types"

const metaOptions = ["id", "comment", "props"]

const escapedMetaOptionsRe = new RegExp(`^_(${metaOptions.join("|")})$`)

export type ParsedResult = {
  message: string
  values?: Record<string, Expression>
  jsxElements?: Record<string, JSXElement>
}

export type TextToken = {
  type: "text"
  value: string
}
export type ArgToken = {
  type: "arg"
  value: Expression
  name?: string

  /**
   * plural
   * select
   * selectordinal
   */
  format?: string
  options?: {
    offset: string
    [icuChoice: string]: string | Tokens
  }
}
export type ElementToken = {
  type: "element"
  value: JSXElement
  name?: string | number
  children?: Token[]
}
export type Tokens = Token | Token[]
export type Token = TextToken | ArgToken | ElementToken

export default class ICUMessageFormat {
  public fromTokens(tokens: Tokens): ParsedResult {
    return (Array.isArray(tokens) ? tokens : [tokens])
      .map((token) => this.processToken(token))
      .filter(Boolean)
      .reduce(
        (props, message) => ({
          ...message,
          message: props.message + message.message,
          values: { ...props.values, ...message.values },
          jsxElements: { ...props.jsxElements, ...message.jsxElements },
        }),
        {
          message: "",
          values: {},
          jsxElements: {},
        }
      )
  }

  public processToken(token: Token): ParsedResult {
    const jsxElements: ParsedResult["jsxElements"] = {}

    if (token.type === "text") {
      return {
        message: token.value,
      }
    } else if (token.type === "arg") {
      if (
        token.value !== undefined &&
        isJSXEmptyExpression(token.value as Node)
      ) {
        return null
      }
      const values =
        token.value !== undefined ? { [token.name]: token.value } : {}

      switch (token.format) {
        case "plural":
        case "select":
        case "selectordinal":
          const formatOptions = Object.keys(token.options)
            .filter((key) => token.options[key] != null)
            .map((key) => {
              let value = token.options[key]
              key = key.replace(escapedMetaOptionsRe, "$1")

              if (key === "offset") {
                // offset has special syntax `offset:number`
                return `offset:${value}`
              }

              if (typeof value !== "string") {
                // process tokens from nested formatters
                const {
                  message,
                  values: childValues,
                  jsxElements: childJsxElements,
                } = this.fromTokens(value)

                Object.assign(values, childValues)
                Object.assign(jsxElements, childJsxElements)
                value = message
              }

              return `${key} {${value}}`
            })
            .join(" ")

          return {
            message: `{${token.name}, ${token.format}, ${formatOptions}}`,
            values,
            jsxElements,
          }
        default:
          return {
            message: `{${token.name}}`,
            values,
          }
      }
    } else if (token.type === "element") {
      let message = ""
      let elementValues: ParsedResult["values"] = {}
      Object.assign(jsxElements, { [token.name]: token.value })
      token.children.forEach((child) => {
        const {
          message: childMessage,
          values: childValues,
          jsxElements: childJsxElements,
        } = this.fromTokens(child)

        message += childMessage
        Object.assign(elementValues, childValues)
        Object.assign(jsxElements, childJsxElements)
      })
      return {
        message: token.children.length
          ? `<${token.name}>${message}</${token.name}>`
          : `<${token.name}/>`,
        values: elementValues,
        jsxElements,
      }
    }

    throw new Error(`Unknown token type ${(token as any).type}`)
  }
}
