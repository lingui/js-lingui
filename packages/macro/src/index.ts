import { createMacro, MacroParams } from "babel-plugin-macros"
import { getConfig as loadConfig, LinguiConfigNormalized } from "@lingui/conf"

import MacroJS from "./macroJs"
import MacroJSX from "./macroJsx"
import { NodePath } from "@babel/traverse"
import {
  ImportDeclaration,
  Identifier,
  isImportSpecifier,
  isIdentifier,
  JSXIdentifier,
} from "@babel/types"

export type LinguiMacroOpts = {
  // explicitly set by CLI when running extraction process
  extract?: boolean
  linguiConfig?: LinguiConfigNormalized
}

const jsMacroTags = new Set([
  "defineMessage",
  "msg",
  "arg",
  "t",
  "plural",
  "select",
  "selectOrdinal",
])

const jsxMacroTags = new Set(["Trans", "Plural", "Select", "SelectOrdinal"])

let config: LinguiConfigNormalized

function getConfig(_config?: LinguiConfigNormalized) {
  if (_config) {
    config = _config
  }
  if (!config) {
    config = loadConfig()
  }
  return config
}

function macro({ references, state, babel, config }: MacroParams) {
  const opts: LinguiMacroOpts = config as LinguiMacroOpts

  const {
    i18nImportModule,
    i18nImportName,
    TransImportModule,
    TransImportName,
  } = getConfig(opts.linguiConfig).runtimeConfigModule

  const jsxNodes = new Set<NodePath>()
  const jsNodes = new Set<NodePath>()
  let needsI18nImport = false

  let nameMap = new Map<string, string>()
  Object.keys(references).forEach((tagName) => {
    const nodes = references[tagName]

    if (jsMacroTags.has(tagName)) {
      nodes.forEach((path) => {
        nameMap.set(tagName, (path.node as Identifier).name)
        jsNodes.add(path.parentPath)
      })
    } else if (jsxMacroTags.has(tagName)) {
      // babel-plugin-macros return JSXIdentifier nodes.
      // Which is for every JSX element would be presented twice (opening / close)
      // Here we're taking JSXElement and dedupe it.
      nodes.forEach((path) => {
        nameMap.set(tagName, (path.node as JSXIdentifier).name)

        // identifier.openingElement.jsxElement
        jsxNodes.add(path.parentPath.parentPath)
      })
    } else {
      throw nodes[0].buildCodeFrameError(`Unknown macro ${tagName}`)
    }
  })

  const stripNonEssentialProps =
    process.env.NODE_ENV == "production" && !opts.extract

  const jsNodesArray = Array.from(jsNodes)

  jsNodesArray.filter(isRootPath(jsNodesArray)).forEach((path) => {
    const macro = new MacroJS(babel, {
      i18nImportName,
      stripNonEssentialProps,
      nameMap,
    })
    try {
      if (macro.replacePath(path)) needsI18nImport = true
    } catch (e) {
      reportUnsupportedSyntax(path, e as Error)
    }
  })

  const jsxNodesArray = Array.from(jsxNodes)

  jsxNodesArray.filter(isRootPath(jsxNodesArray)).forEach((path) => {
    const macro = new MacroJSX(babel, { stripNonEssentialProps, nameMap })

    try {
      macro.replacePath(path)
    } catch (e) {
      reportUnsupportedSyntax(path, e as Error)
    }
  })

  if (needsI18nImport) {
    addImport(babel, state, i18nImportModule, i18nImportName)
  }

  if (jsxNodes.size) {
    addImport(babel, state, TransImportModule, TransImportName)
  }
}

function reportUnsupportedSyntax(path: NodePath, e: Error) {
  throw path.buildCodeFrameError(
    `Unsupported macro usage. Please check the examples at https://lingui.dev/ref/macro#examples-of-js-macros. 
 If you think this is a bug, fill in an issue at https://github.com/lingui/js-lingui/issues
 
 Error: ${e.message}`
  )
}

function addImport(
  babel: MacroParams["babel"],
  state: MacroParams["state"],
  module: string,
  importName: string
) {
  const { types: t } = babel

  const linguiImport = state.file.path.node.body.find(
    (importNode) =>
      t.isImportDeclaration(importNode) &&
      importNode.source.value === module &&
      // https://github.com/lingui/js-lingui/issues/777
      importNode.importKind !== "type"
  ) as ImportDeclaration

  const tIdentifier = t.identifier(importName)
  // Handle adding the import or altering the existing import
  if (linguiImport) {
    if (
      linguiImport.specifiers.findIndex(
        (specifier) =>
          isImportSpecifier(specifier) &&
          isIdentifier(specifier.imported, { name: importName })
      ) === -1
    ) {
      linguiImport.specifiers.push(t.importSpecifier(tIdentifier, tIdentifier))
    }
  } else {
    state.file.path.node.body.unshift(
      t.importDeclaration(
        [t.importSpecifier(tIdentifier, tIdentifier)],
        t.stringLiteral(module)
      )
    )
  }
}

/**
 * Filtering nested macro calls
 *
 * <Macro>
 *   <Macro /> <-- this would be filtered out
 * </Macro>
 */
function isRootPath(allPath: NodePath[]) {
  return (node: NodePath) =>
    (function traverse(path): boolean {
      if (!path.parentPath) {
        return true
      } else {
        return !allPath.includes(path.parentPath) && traverse(path.parentPath)
      }
    })(node)
}

;[...jsMacroTags, ...jsxMacroTags].forEach((name) => {
  Object.defineProperty(module.exports, name, {
    get() {
      throw new Error(
        `The macro you imported from "@lingui/macro" is being executed outside the context of compilation with babel-plugin-macros. ` +
          `This indicates that you don't have the babel plugin "babel-plugin-macros" configured correctly. ` +
          `Please see the documentation for how to configure babel-plugin-macros properly: ` +
          "https://github.com/kentcdodds/babel-plugin-macros/blob/main/other/docs/user.md"
      )
    },
  })
})

export default createMacro(macro, {
  configName: "lingui",
})
