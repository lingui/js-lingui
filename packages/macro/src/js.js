const nlRe = /(?:\r\n|\r|\n)+\s+/g

const pluralRules = ["zero", "one", "two", "few", "many", "other"]

const generatorFactory = (index = 0) => () => index++

const initialProps = () => ({
  text: "",
  values: {},
  formats: {},
  argumentGenerator: generatorFactory()
})

export default function({ types: t }) {
  const isIdentifier = (node, name) => t.isIdentifier(node, { name })

  const isI18nMethod = node =>
    isIdentifier(node.tag, "t") ||
    (t.isCallExpression(node.tag) && isIdentifier(node.tag.callee, "t"))

  const isChoiceMethod = node =>
    isIdentifier(node, "plural") ||
    isIdentifier(node, "select") ||
    isIdentifier(node, "selectOrdinal")

  const isFormatMethod = node =>
    isIdentifier(node, "date") || isIdentifier(node, "number")

  /**
   * Convert identifiers to *named* arguments and everything else
   * to *positional* arguments.
   *
   * Example
   * `world` is named argument and `new Date()` is positional one
   *
   * Input:   `Hello ${world}, today is ${new Date()}`
   * Output:  `Hello {world}, today is {0}`
   */
  function expressionToArgument(exp, props) {
    const name = t.isIdentifier(exp) ? exp.name : props.argumentGenerator()
    const key = t.isIdentifier(exp) ? exp : t.numericLiteral(name)

    return { name, key }
  }

  function transformI18nMethod(node, file, props) {
    if (t.isCallExpression(node.tag)) {
      // Message with custom ID, where message is used as defaults
      // i18n.t('id')`Hello World`
      const defaults = node.tag.arguments[0]
      if (!t.isStringLiteral(defaults)) {
        throw file.buildCodeFrameError(node.tag, "Message ID must be a string")
      }
      const newProps = transformTemplateLiteral(node.quasi, file, props)
      return {
        ...newProps,
        text: defaults.value,
        defaults: props.text
      }
    }

    return transformTemplateLiteral(node.quasi, file, props)
  }

  function transformChoiceMethod(node, file, props, root = false) {
    const choices = {}
    const choicesType = node.callee.name.toLowerCase()
    let defaults
    let variable
    let offset = ""

    let choiceArguments = node.arguments[0]
    if (t.isStringLiteral(choiceArguments)) {
      defaults = choiceArguments.value
      choiceArguments = node.arguments[1]
    }

    for (const attr of choiceArguments.properties) {
      if (attr.computed) {
        throw file.buildCodeFrameError(
          attr,
          "Computed properties aren't allowed."
        )
      }

      const { key } = attr
      // key is either:
      // NumericLiteral => convert to `={number}`
      // StringLiteral => key.value
      // Literal => key.name
      const name = t.isNumericLiteral(key)
        ? `=${key.value}`
        : key.name || key.value

      if (name === "value") {
        const exp = attr.value
        const { name, key } = expressionToArgument(exp, props)
        variable = name
        props.values[name] = t.objectProperty(key, exp)
      } else if (choicesType !== "select" && name === "offset") {
        // offset is static parameter, so it must be either string or number
        if (!t.isNumericLiteral(attr.value) && !t.isStringLiteral(attr.value)) {
          throw file.buildCodeFrameError(
            node.callee,
            "Offset argument cannot be a variable."
          )
        }
        offset = ` offset:${attr.value.value}`
      } else {
        let value = ""

        if (t.isTemplateLiteral(attr.value)) {
          props = transformTemplateLiteral(attr.value, file, {
            ...props,
            text: ""
          })
          value = props.text
        } else if (t.isCallExpression(attr.value)) {
          props = transformMethod(attr.value, file, { ...props, text: "" })
          value = `{${props.text}}`
        } else {
          value = attr.value.value
        }
        choices[name] = value
      }
    }

    // missing value
    if (variable === undefined) {
      throw file.buildCodeFrameError(node.callee, "Value argument is missing.")
    }

    const choicesKeys = Object.keys(choices)

    // 'other' choice is required
    if (!choicesKeys.length) {
      throw file.buildCodeFrameError(
        node.callee,
        `Missing ${choicesType} choices. At least fallback argument 'other' is required.`
      )
    } else if (!Array.includes(choicesKeys, "other")) {
      throw file.buildCodeFrameError(
        node.callee,
        `Missing fallback argument 'other'.`
      )
    }

    // validate plural rules
    if (choicesType === "plural" || choicesType === "selectordinal") {
      choicesKeys.forEach(rule => {
        if (!Array.includes(pluralRules, rule) && !/=\d+/.test(rule)) {
          throw file.buildCodeFrameError(
            node.callee,
            `Invalid plural rule '${rule}'. Must be ${pluralRules.join(
              ", "
            )} or exact number depending on your source language ('one' and 'other' for English).`
          )
        }
      })
    }

    const argument = choicesKeys
      .map(form => `${form} {${choices[form]}}`)
      .join(" ")
    const format = `${variable}, ${choicesType},${offset} ${argument}`
    props.text = root ? `{${format}}` : format

    if (defaults) {
      return {
        ...props,
        text: defaults,
        defaults: props.text
      }
    }

    return props
  }

  function transformFormatMethod(node, file, props, root) {
    const exp = node.arguments[0]

    // missing value
    if (exp === undefined) {
      throw file.buildCodeFrameError(
        node.callee,
        "The first argument of format function must be a variable."
      )
    }

    const { name, key } = expressionToArgument(exp, props)

    const type = node.callee.name
    const parts = [
      name, // variable name
      type // format type
    ]

    let format = ""
    const formatArg = node.arguments[1]
    if (!formatArg) {
      // Do not throw validation error when format doesn't exist
    } else if (t.isStringLiteral(formatArg)) {
      format = formatArg.value
    } else if (t.isIdentifier(formatArg) || t.isObjectExpression(formatArg)) {
      if (t.isIdentifier(formatArg)) {
        format = formatArg.name
      } else {
        const formatName = new RegExp(`^${type}\\d+$`)
        const existing = Object.keys(props.formats).filter(name =>
          formatName.test(name)
        )
        format = `${type}${existing.length || 0}`
      }

      props.formats[format] = t.objectProperty(t.identifier(format), formatArg)
    } else {
      throw file.buildCodeFrameError(
        formatArg,
        "Format can be either string for buil-in formats, variable or object for custom defined formats."
      )
    }

    if (format) parts.push(format)

    props.values[name] = t.objectProperty(key, exp)
    props.text += `${parts.join(",")}`

    return props
  }

  function transformTemplateLiteral(exp, file, props) {
    let parts = []

    exp.quasis.forEach((item, index) => {
      parts.push(item)

      if (!item.tail) parts.push(exp.expressions[index])
    })

    parts.forEach(item => {
      if (t.isTemplateElement(item)) {
        props.text += item.value.raw.replace(/\\`/g, "`")
      } else if (
        t.isCallExpression(item) &&
        (isI18nMethod(item.callee) ||
          isChoiceMethod(item.callee) ||
          isFormatMethod(item.callee))
      ) {
        const { text } = transformMethod(item, file, {
          ...props,
          text: ""
        })
        props.text += `{${text}}`
      } else {
        const { name, key } = expressionToArgument(item, props)

        props.text += `{${name}}`
        props.values[name] = t.objectProperty(key, item)
      }
    })

    return props
  }

  function transformMethod(node, file, props, root = false) {
    if (isI18nMethod(node)) {
      // t
      return transformI18nMethod(node, file, props)
    } else if (isChoiceMethod(node.callee)) {
      // plural, select and selectOrdinal
      return transformChoiceMethod(node, file, props, root)
    } else if (isFormatMethod(node.callee)) {
      // date, number
      return transformFormatMethod(node, file, props, root)
    }

    return props
  }

  return function transform(path, file) {
    // 1. Collect all parameters and generate message ID
    const props = transformMethod(path.node, file, initialProps(), true)

    const text = props.text.replace(nlRe, " ").trim()
    if (!text) return

    // 2. Create message descriptor
    const descriptorProps = [
      t.objectProperty(t.identifier("id"), t.StringLiteral(text))
    ]

    if (props.defaults) {
      descriptorProps.push(
        t.objectProperty(
          t.identifier("defaults"),
          t.stringLiteral(props.defaults)
        )
      )
    }

    const formatsList = Object.values(props.formats)
    if (formatsList.length) {
      descriptorProps.push(
        t.objectProperty(
          t.identifier("formats"),
          t.objectExpression(formatsList)
        )
      )
    }

    const valuesList = Object.values(props.values)
    if (valuesList.length) {
      descriptorProps.push(
        t.objectProperty(
          t.identifier("values"),
          t.objectExpression(valuesList.length ? valuesList : [])
        )
      )
    }

    const exp = t.objectExpression(descriptorProps)
    exp.loc = path.node.loc
    path.replaceWith(exp)

    // 3. Add leading `i18n` comment (if doesn't exist) for lingui-extract
    let nodeWithComments =
      path.parentPath.type === "ExpressionStatement"
        ? path.parentPath.node
        : path.node
    const i18nComment =
      nodeWithComments.leadingComments &&
      nodeWithComments.leadingComments.filter(
        node => node.value.trim().substr(0, 4) === "i18n"
      )[0]

    if (!i18nComment) {
      path.addComment("leading", "i18n")
    }
  }
}
