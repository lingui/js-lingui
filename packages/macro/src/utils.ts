import * as R from "ramda"

/**
 * Custom zip method which takes length of the larger array
 * (usually zip functions use the `smaller` length, discarding values in larger array)
 */
export function zip<A, B>(a: A[], b: B[]): [A, B][] {
  return R.range(0, Math.max(a.length, b.length)).map((index) => [
    a[index],
    b[index],
  ])
}

export const makeCounter = (index = 0) => () => index++
