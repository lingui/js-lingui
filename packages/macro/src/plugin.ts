import type { PluginObj, Visitor, PluginPass } from "@babel/core"
import type * as babelTypes from "@babel/types"
import MacroJSX from "./macroJsx"
import { NodePath } from "@babel/traverse"
import MacroJs from "./macroJs"
import { MACRO_PACKAGE } from "./constants"
import {
  type LinguiConfigNormalized,
  getConfig as loadConfig,
} from "@lingui/conf"
import { Program, Identifier } from "@babel/types"

let config: LinguiConfigNormalized

export type LinguiPluginOpts = {
  // explicitly set by CLI when running extraction process
  extract?: boolean
  linguiConfig?: LinguiConfigNormalized
}

function getConfig(_config?: LinguiConfigNormalized) {
  if (_config) {
    config = _config
  }
  if (!config) {
    config = loadConfig()
  }
  return config
}

function reportUnsupportedSyntax(path: NodePath, e: Error) {
  const codeFrameError = path.buildCodeFrameError(
    `Unsupported macro usage. Please check the examples at https://lingui.dev/ref/macro#examples-of-js-macros.
 If you think this is a bug, fill in an issue at https://github.com/lingui/js-lingui/issues

 Error: ${e.message}`
  )

  // show stack trace where error originally happened
  codeFrameError.stack = e.stack
  throw codeFrameError
}

type LinguiSymbol = "Trans" | "useLingui" | "i18n"

export default function ({
  types: t,
}: {
  types: typeof babelTypes
}): PluginObj {
  function addImport(state: PluginPass, name: LinguiSymbol) {
    const path = state.get(
      "macroImport"
    ) as NodePath<babelTypes.ImportDeclaration>

    const config = state.get("linguiConfig") as LinguiConfigNormalized

    if (!state.get("has_import_" + name)) {
      state.set("has_import_" + name, true)
      const [moduleSource, importName] = config.runtimeConfigModule[name]

      const [newPath] = path.insertAfter(
        t.importDeclaration(
          [
            t.importSpecifier(
              getSymbolIdentifier(state, name),
              t.identifier(importName)
            ),
          ],
          t.stringLiteral(moduleSource)
        )
      )

      path.parentPath.scope.registerDeclaration(newPath)
    }

    return path.parentPath.scope.getBinding(
      getSymbolIdentifier(state, name).name
    )
  }

  function getMacroImports(path: NodePath<Program>) {
    return path.get("body").filter((statement) => {
      return (
        statement.isImportDeclaration() &&
        statement.get("source").node.value === MACRO_PACKAGE
      )
    })
  }

  function getSymbolIdentifier(
    state: PluginPass,
    name: LinguiSymbol
  ): Identifier {
    return state.get("linguiIdentifiers")[name]
  }

  return {
    name: "lingui-macro-plugin",
    visitor: {
      Program: {
        enter(path, state) {
          const macroImports = getMacroImports(path)

          if (!macroImports.length) {
            return
          }

          state.set("macroImport", macroImports[0])

          state.set(
            "linguiConfig",
            getConfig((state.opts as LinguiPluginOpts).linguiConfig)
          )

          state.set("linguiIdentifiers", {
            i18n: path.scope.generateUidIdentifier("i18n"),
            Trans: path.scope.generateUidIdentifier("Trans"),
            useLingui: path.scope.generateUidIdentifier("useLingui"),
          })

          path.traverse(
            {
              JSXElement(path, state) {
                const macro = new MacroJSX(
                  { types: t },
                  {
                    transImportName: getSymbolIdentifier(state, "Trans").name,
                    stripNonEssentialProps:
                      process.env.NODE_ENV == "production" &&
                      !(state.opts as LinguiPluginOpts).extract,
                  }
                )

                let newNode: false | babelTypes.Node

                try {
                  newNode = macro.replacePath(path)
                } catch (e) {
                  reportUnsupportedSyntax(path, e as Error)
                }

                if (newNode) {
                  const [newPath] = path.replaceWith(newNode)
                  addImport(state, "Trans").reference(newPath)
                }
              },

              "CallExpression|TaggedTemplateExpression"(
                path: NodePath<
                  | babelTypes.CallExpression
                  | babelTypes.TaggedTemplateExpression
                >,
                state: PluginPass
              ) {
                const macro = new MacroJs(
                  { types: t },
                  {
                    stripNonEssentialProps:
                      process.env.NODE_ENV == "production" &&
                      !(state.opts as LinguiPluginOpts).extract,
                    i18nImportName: getSymbolIdentifier(state, "i18n").name,
                    useLinguiImportName: getSymbolIdentifier(state, "useLingui")
                      .name,
                  }
                )
                let newNode: false | babelTypes.Node

                try {
                  newNode = macro.replacePath(path)
                } catch (e) {
                  reportUnsupportedSyntax(path, e as Error)
                }

                if (newNode) {
                  const [newPath] = path.replaceWith(newNode)

                  if (macro.needsUseLinguiImport) {
                    addImport(state, "useLingui").reference(newPath)
                  }

                  if (macro.needsI18nImport) {
                    addImport(state, "i18n").reference(newPath)
                  }
                }
              },
            } as Visitor<PluginPass>,
            state
          )
        },
        exit(path, state) {
          const macroImports = getMacroImports(path)
          macroImports.forEach((path) => path.remove())
        },
      },
    } as Visitor<PluginPass>,
  }
}
