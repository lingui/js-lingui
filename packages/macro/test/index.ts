import fs from "fs"
import path from "path"
import {
  PluginObj,
  transformFileSync,
  TransformOptions,
  transformSync,
} from "@babel/core"
import prettier from "prettier"
import { LinguiMacroOpts } from "../src/index"
import {
  JSXAttribute,
  jsxExpressionContainer,
  JSXIdentifier,
  stringLiteral,
} from "@babel/types"
import { NodePath } from "@babel/traverse"

export type TestCase = {
  name?: string
  input?: string
  expected?: string
  filename?: string
  production?: boolean
  useTypescriptPreset?: boolean
  macroOpts?: LinguiMacroOpts
  /** Remove hash id from snapshot for more stable testing */
  stripId?: boolean
  only?: boolean
  skip?: boolean
}

const testCases: Record<string, TestCase[]> = {
  "js-t": require("./js-t").default,
  "js-plural": require("./js-plural").default,
  "js-select": require("./js-select").default,
  "js-selectOrdinal": require("./js-selectOrdinal").default,
  "jsx-trans": require("./jsx-trans").default,
  "jsx-select": require("./jsx-select").default,
  "jsx-plural": require("./jsx-plural").default,
  "jsx-selectOrdinal": require("./jsx-selectOrdinal").default,
  "js-defineMessage": require("./js-defineMessage").default,
}

function stripIdPlugin(): PluginObj {
  return {
    visitor: {
      JSXOpeningElement: (path) => {
        const idAttr = path
          .get("attributes")
          .find(
            (attr) =>
              attr.isJSXAttribute() &&
              (attr.node.name as JSXIdentifier).name === "id"
          ) as NodePath<JSXAttribute>

        if (idAttr) {
          idAttr
            .get("value")
            .replaceWith(jsxExpressionContainer(stringLiteral("<stripped>")))
        }
      },
    },
  }
}

describe("macro", function () {
  process.env.LINGUI_CONFIG = path.join(__dirname, "lingui.config.js")

  const getDefaultBabelOptions = (
    macroOpts: LinguiMacroOpts = {},
    isTs: boolean = false,
    stripId = false
  ): TransformOptions => {
    return {
      filename: "<filename>" + (isTs ? ".tsx" : "jsx"),
      configFile: false,
      presets: [],
      plugins: [
        "@babel/plugin-syntax-jsx",
        [
          "macros",
          {
            lingui: macroOpts,
            // macro plugin uses package `resolve` to find a path of macro file
            // this will not follow jest pathMapping and will resolve path from ./build
            // instead of ./src which makes testing & developing hard.
            // here we override resolve and provide correct path for testing
            resolvePath: (source: string) => require.resolve(source),
          },
        ],
        ...(stripId ? [stripIdPlugin] : []),
      ],
    }
  }

  // return function, so we can test exceptions
  const transformCode = (code: string) => () => {
    try {
      return transformSync(code, getDefaultBabelOptions()).code.trim()
    } catch (e) {
      ;(e as Error).message = (e as Error).message.replace(/([^:]*:){2}/, "")
      throw e
    }
  }

  Object.keys(testCases).forEach((suiteName) => {
    describe(suiteName, () => {
      const cases = testCases[suiteName]

      const clean = (value: string) =>
        prettier.format(value, { parser: "babel-ts" }).replace(/\n+/, "\n")

      cases.forEach(
        (
          {
            name,
            input,
            expected,
            filename,
            production,
            useTypescriptPreset,
            only,
            skip,
            macroOpts,
            stripId,
          },
          index
        ) => {
          let run = it
          if (only) run = it.only
          if (skip) run = it.skip
          run(name != null ? name : `${suiteName} #${index + 1}`, () => {
            const babelOptions = getDefaultBabelOptions(
              macroOpts,
              useTypescriptPreset,
              stripId
            )
            expect(input || filename).toBeDefined()

            const originalEnv = process.env.NODE_ENV

            if (production) {
              process.env.NODE_ENV = "production"
            }

            if (useTypescriptPreset) {
              babelOptions.presets.push("@babel/preset-typescript")
            }

            try {
              if (filename) {
                const inputPath = path.relative(
                  process.cwd(),
                  path.join(__dirname, "fixtures", filename)
                )
                const expectedPath = inputPath.replace(/\.js$/, ".expected.js")
                const expected = fs
                  .readFileSync(expectedPath, "utf8")
                  .replace(/\r/g, "")
                  .trim()

                const _babelOptions = {
                  ...babelOptions,
                  cwd: path.dirname(inputPath),
                }

                const actual = transformFileSync(inputPath, _babelOptions)
                  .code.replace(/\r/g, "")
                  .trim()
                expect(clean(actual)).toEqual(clean(expected))
              } else {
                const actual = transformSync(input, babelOptions).code.trim()

                expect(clean(actual)).toEqual(clean(expected))
              }
            } finally {
              process.env.LINGUI_CONFIG = ""
              process.env.NODE_ENV = originalEnv
            }
          })
        }
      )
    })
  })

  it("Should throw error if used without babel-macro-plugin", async () => {
    await expect(async () => {
      const mod = await import("../src/index")
      return (mod as unknown as typeof import("@lingui/macro")).Trans
    }).rejects.toThrow('The macro you imported from "@lingui/macro"')
  })

  describe.skip("validation", function () {
    describe("plural/select/selectordinal", function () {
      it("value is missing", function () {
        const code = `
        plural({
          0: "No books",
          1: "1 book",
          other: "# books"
        });`
        expect(transformCode(code)).toThrowErrorMatchingSnapshot()
      })

      it("offset must be number or string, not variable", function () {
        const code = `
        plural({
          offset: count,
          0: "No books",
          1: "1 book",
          other: "# books"
        });`
        expect(transformCode(code)).toThrowErrorMatchingSnapshot()
      })

      it("plural forms are missing", function () {
        const plural = `
        plural({
          value: count
        });`
        expect(transformCode(plural)).toThrowErrorMatchingSnapshot()

        const select = `
        plural({
          value: count
        });`
        expect(transformCode(select)).toThrowErrorMatchingSnapshot()

        const selectOrdinal = `
        plural({
          value: count
        });`
        expect(transformCode(selectOrdinal)).toThrowErrorMatchingSnapshot()
      })

      it("plural forms cannot be variables", function () {
        const code = `
        plural({
          value: count,
          [one]: "Book"
        });`
        expect(transformCode(code)).toThrowErrorMatchingSnapshot()
      })

      it("plural rules must be valid", function () {
        const plural = `
        plural({
          value: count,
          one: "Book",
          three: "Invalid",
          other: "Books"
        });`
        expect(transformCode(plural)).toThrowErrorMatchingSnapshot()

        const selectOrdinal = `
        selectOrdinal({
          value: count,
          one: "st",
          three: "Invalid",
          other: "rd"
        });`
        expect(transformCode(selectOrdinal)).toThrowErrorMatchingSnapshot()
      })
    })

    describe("formats", function () {
      it("value is missing", function () {
        expect(transformCode("date();")).toThrowErrorMatchingSnapshot()
      })

      it("format must be either string, variable or object with custom format", function () {
        expect(transformCode('number(value, "currency");')).not.toThrow()
        expect(transformCode("number(value, currency);")).not.toThrow()
        expect(transformCode("number(value, { digits: 4 });")).not.toThrow()
        expect(
          transformCode("number(value, 42);")
        ).toThrowErrorMatchingSnapshot()
      })
    })

    describe("Plural/Select/SelectOrdinal", function () {
      it("children are not allowed", function () {
        expect(
          transformCode("<Plural>Not allowed</Plural>")
        ).toThrowErrorMatchingSnapshot()
        expect(
          transformCode("<Select>Not allowed</Select>")
        ).toThrowErrorMatchingSnapshot()
        expect(
          transformCode("<SelectOrdinal>Not allowed</SelectOrdinal>")
        ).toThrowErrorMatchingSnapshot()
      })

      it("value is missing", function () {
        const code = `<Plural one="Book" other="Books" />`
        expect(transformCode(code)).toThrowErrorMatchingSnapshot()
      })

      it("offset must be number or string, not variable", function () {
        const variable = `<Plural value={value} offset={offset} one="Book" other="Books" />`
        expect(transformCode(variable)).toThrowErrorMatchingSnapshot()
      })

      it("plural forms are missing", function () {
        const plural = `<Plural value={value} />`
        expect(transformCode(plural)).toThrowErrorMatchingSnapshot()

        const select = `<Select value={value} />`
        expect(transformCode(select)).toThrowErrorMatchingSnapshot()

        const ordinal = `<SelectOrdinal value={value} />`
        expect(transformCode(ordinal)).toThrowErrorMatchingSnapshot()
      })

      it("plural rules must be valid", function () {
        const plural = `<Plural value={value} three="Invalid" one="Book" other="Books" />`
        expect(transformCode(plural)).toThrowErrorMatchingSnapshot()

        const ordinal = `<SelectOrdinal value={value} three="Invalid" one="st" other="rd" />`
        expect(transformCode(ordinal)).toThrowErrorMatchingSnapshot()
      })
    })
  })
})
