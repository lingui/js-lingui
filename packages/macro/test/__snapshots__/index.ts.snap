// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macro useLingui Should throw if not used with destructuring 1`] = `
 Unsupported macro usage. Please check the examples at https://lingui.dev/ref/macro#examples-of-js-macros.
 If you think this is a bug, fill in an issue at https://github.com/lingui/js-lingui/issues

 Error: Must destruct \`t\` when using \`useLingui\` macro, i.e:
 const { t } = useLingui()
 or
 const { t: _ } = useLingui()
 
  2 |       import {useLingui} from "@lingui/macro";
  3 |       
> 4 |       const lingui = useLingui()
    |             ^^^^^^^^^^^^^^^^^^^^
  5 |        
  6 |         Learn more: https://www.npmjs.com/package/@lingui/macro
`;

exports[`macro useLingui Should throw if used not in the variable declaration 1`] = `
 Unsupported macro usage. Please check the examples at https://lingui.dev/ref/macro#examples-of-js-macros.
 If you think this is a bug, fill in an issue at https://github.com/lingui/js-lingui/issues

 Error: \`useLingui\` macro must be used in variable declaration.
        
 Example:

 const { t } = useLingui()

  2 |       import {useLingui} from "@lingui/macro";
  3 |       
> 4 |       useLingui()
    |       ^^^^^^^^^
  5 |        
  6 |         Learn more: https://www.npmjs.com/package/@lingui/macro
`;
